BUG: Breaking change to offsets.pyx - quarter offset classes

Pawel-Kranzberg opened the pull request:
Set default months for `BQuarterEnd`, `BQuarterBegin`, `QuarterEnd`, `QuarterBegin` classes in order; rename `startingMonth` parameter to `month` for those classes.
Addresses #8435 and #5307

- [ ] closes #8435 #5307
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Pawel-Kranzberg committed:
Breaking change to offsets.pyx

Addresses #8435

jreback commented:
pls just one PR - just push to the same one if changing

Pawel-Kranzberg commented:
> pls just one PR - just push to the same one if changing

But this one has different scope from #39887 .

jreback commented:
ok that's fine

you always need a test

jreback reviewed (changes requested):
you would need to change / add tests to show what is changing. and have the tests pass

jbrockmendel commented:
does this handle #5307?

Pawel-Kranzberg commented:
> does this handle #5307?

Not at present. I'd like to cover it as well - please let me know if that's OK.

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg commented:
> does this handle #5307?

It does now.

Pawel-Kranzberg committed:
Update test_constructors.py

Pawel-Kranzberg committed:
Update test_misc.py

Pawel-Kranzberg committed:
Update test_period_index.py

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update test_dst.py

Pawel-Kranzberg committed:
Update test_dst.py

Pawel-Kranzberg committed:
Update test_offsets.py

Pawel-Kranzberg committed:
Update test_offsets.py

Pawel-Kranzberg committed:
Update test_constructors.py

Pawel-Kranzberg committed:
Update test_misc.py

Pawel-Kranzberg committed:
Update test_offsets.py

Pawel-Kranzberg committed:
Update test_yqm_offsets.py

Pawel-Kranzberg committed:
Update test_period_index.py

Pawel-Kranzberg committed:
Update test_offsets.py

Pawel-Kranzberg committed:
Update test_offsets.py

jreback reviewed (changes requested):
so you are adding a new parameter 'month', so now we have the confusing 'startingMonth' and 'month'. I would rather deprecate 'startingMonth', then add new; not even sure 'month' is a good name here. can you write a sample whatsnew note indicate what the change is here from a user POV.

Pawel-Kranzberg committed:
Rename startingMonth to month

Rename `startingMonth` to `month` for `BQuarterEnd`, `BQuarterBegin`, `QuarterEnd`, `QuarterBegin`.

Pawel-Kranzberg committed:
Merge branch 'master' into patch-2

Pawel-Kranzberg committed:
Update test_to_offset.py

Rename startingMonth to month.

Pawel-Kranzberg committed:
Update test_libfrequencies.py

Rename startingMonth to month.

Pawel-Kranzberg committed:
Update test_period.py

Rename startingMonth to month

Pawel-Kranzberg committed:
Update offsets.pyx

Pickling error investigation.

Pawel-Kranzberg committed:
Remove startingMonth reference

Remove startingMonth reference from test_basic_downsample() in test_period_index.py

Pawel-Kranzberg committed:
Rename startingMonth to month

Rename startingMonth to month in test_period.py

Pawel-Kranzberg committed:
Update offset_frequency.rst

Rename startingMonth to month.

Pawel-Kranzberg committed:
DOC: Rename startingMonth to month

Pawel-Kranzberg committed:
Rename startingMonth to month

Pawel-Kranzberg committed:
Fix "QuarterBegin" timestamp in test_rollback

Pawel-Kranzberg committed:
Fix timestamp for QuarterBegin

Pawel-Kranzberg committed:
Fix offsets tests for QuarterBegin

Pawel-Kranzberg committed:
Rename startingMonth to month

Pawel-Kranzberg committed:
Fix pickling of quarter-based offsets

Pawel-Kranzberg committed:
Fix pickling of quarter-based offsets

Pawel-Kranzberg committed:
Deprecate startingMonth

Pawel-Kranzberg committed:
Merge pull request #3 from pandas-dev/master

Sync with master

Pawel-Kranzberg committed:
Add deprecation warning test

Pawel-Kranzberg committed:
Add test_startingMonth_deprecation_warning

Pawel-Kranzberg committed:
Add startingMonth deprecation to QuarterEnd

pep8speaks commented:
Hello @Pawel-Kranzberg! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:















* In the file [`pandas/tests/tseries/offsets/test_offsets.py`](https://github.com/pandas-dev/pandas/blob/57fe35049f0662a52050daa600ad3e70d0bc8847/pandas/tests/tseries/offsets/test_offsets.py):

> [Line 876:1](https://github.com/pandas-dev/pandas/blob/57fe35049f0662a52050daa600ad3e70d0bc8847/pandas/tests/tseries/offsets/test_offsets.py#L876): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Add startingMonth deprecation to QuarterEnd

Pawel-Kranzberg committed:
startingMonth depreciation fix

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Create QuarterBase docstring

Pawel-Kranzberg committed:
Update docstrings for quarter-based offsets

Pawel-Kranzberg committed:
Update v0.18.0.rst

Pawel-Kranzberg committed:
Update offset_frequency.rst

Pawel-Kranzberg committed:
Update docstrings

Pawel-Kranzberg committed:
whatsnew entry

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1840,7 +1840,7 @@ cdef class YearOffset(SingleConstructorOffset):
     _attributes = tuple(["n", "normalize", "month"])
 
-    # _default_month: int  # FIXME: python annotation here breaks things
+    # FIXME: python annotation here breaks things
```
do we know what thi refers?

jbrockmendel replied:
putting a python-style annotation `_default_month: int` here led to (i think) compile-time errors

Pawel-Kranzberg replied:
Yes, that's right. I rolled back this change for now, as this part doesn't concern quarters. If OK, I hope to refactor/combine the  QuarterOffset and YearOffset and deal with it at some point.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
+        The calendar month number (1 for January) of the beginning or ending
+        month in a custom-defined quarter to be used as an offset.
+
+        .. versionadded:: 1.X.X
```
1.3

Pawel-Kranzberg replied:
Done.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
-    # FIXME: python annotations here breaks things
-    # _default_starting_month: int
-    # _from_name_starting_month: int
+    # FIXME: python annotation here breaks things
```
can you be more specific

Pawel-Kranzberg replied:
I removed that comment, as other variables are now sufficiently typed - month in particular. Non-ints as default values would be captured by month.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
     cdef readonly:
-        int startingMonth
+        int month
+        int startingMonth  # Backwards compatibility
```
why is this needed? (e.g. can you simply assign it to month when needed?)

Pawel-Kranzberg replied:
You are right. Now I left just the month declaration.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
+        int startingMonth  # Backwards compatibility
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n=1, normalize=False, month=None,
```
can you type

Pawel-Kranzberg replied:
Done.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
-        if startingMonth is None:
-            startingMonth = self._default_starting_month
-        self.startingMonth = startingMonth
+        # Backwards compatibility
```
add the issue number here as a comment

Pawel-Kranzberg replied:
Done.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
-        self.startingMonth = startingMonth
+        # Backwards compatibility
+        if month and startingMonth and month != startingMonth:
+            raise TypeError("Offset received both month and startingMonth")
```
is this line tested?

Pawel-Kranzberg replied:
Thank you, I missed that one. Now it is tested.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
+        elif month is None and startingMonth is not None:
+            warnings.warn(
+                "startingMonth is deprecated, use month instead.",
+                DeprecationWarning,
```
FutureWarning

Pawel-Kranzberg replied:
Done.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
-        self.startingMonth = state.pop("startingMonth")
+        try:
+            self.month = state.pop("month")
+        except:
```
can you be more specifi on the except

Pawel-Kranzberg replied:
I added a longer comment there.

jbrockmendel replied:
a) im pretty sure jreback meant not to use a bare "except:", but to specifically catch `KeyError`

b) rule of thumb: pls let the person who asked the question hit the "resolve conversation" button; something might be clear to you but not to them

jreback replied:
yeah this is non-obvious on the exception.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -2092,26 +2148,44 @@ cdef class BQuarterBegin(QuarterOffset):
+    month : int, default 1
+        The calendar month number (1 for January) of the first month
+        in a custom-defined business quarter to be used as an offset.
+
```
1.3 for here and others

Pawel-Kranzberg replied:
Done.

jreback commented on pandas/tests/tseries/offsets/test_offsets.py:
```diff
@@ -870,3 +873,17 @@ def test_dateoffset_immutable(attribute):
+        "BQuarterEnd",
+    ],
+)
+def test_startingMonth_deprecation_warning(offset_type):
```
FutureWarning and add the issue number as a comment

Pawel-Kranzberg replied:
Done.

jreback reviewed (changes requested):
can you add a sub-section in the whatsnew 1.3 api breaking changes to indicate what a user would except (aside from the deprection)

Pawel-Kranzberg committed:
Update offsets.pyx according to GH#39890 review

Pawel-Kranzberg committed:
Update test_offsets.pyx according to GH#39890 review

Update test_startingMonth_deprecation

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update v1.3.0.rst

Add quarter-based offsets' changes to the API breaking changes (Notable bug fixes) section

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update v1.3.0.rst

Pawel-Kranzberg committed:
Update offset_frequency.rst

Pawel-Kranzberg committed:
Update v0.18.0.rst

Pawel-Kranzberg committed:
Update v1.3.0.rst

Pawel-Kranzberg committed:
Update v0.18.0.rst

Pawel-Kranzberg commented:
> can you add a sub-section in the whatsnew 1.3 api breaking changes to indicate what a user would except (aside from the deprection)

Description drafted and added.

Pawel-Kranzberg commented:
> so you are adding a new parameter 'month', so now we have the confusing 'startingMonth' and 'month'. I would rather deprecate 'startingMonth', then add new; not even sure 'month' is a good name here. can you write a sample whatsnew note indicate what the change is here from a user POV.

BTW, maybe `anchor` would be a better replacement name?

Pawel-Kranzberg committed:
Merge branch 'master' into patch-2

jbrockmendel commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
+Quarter-based time offset classes :class:`~pandas.tseries.offsets.QuarterEnd`,
+:class:`~pandas.tseries.offsets.QuarterBegin`, :class:`~pandas.tseries.offsets.BQuarterEnd`,
+:class:`~pandas.tseries.offsets.BQuarterBegin` now use ``month`` instead of ``startingMonth``
+as a time anchor parameter (:issue:`5307`). Its default value for :class:`~pandas.tseries.offsets.QuarterBegin`
```
"Its" -> "The"?

Pawel-Kranzberg replied:
Good catch, fixed.

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n: int=1, normalize: bool=False, month: int=None,
+                 *, startingMonth: int=None):  # GH#5307 backwards compatibility
```
comment on new line `# GH#5307 startingMonth retained for backwards compatibility`

Pawel-Kranzberg replied:
Comment added.

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
+        int month
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n: int=1, normalize: bool=False, month: int=None,
```
should we try to detect if month is passed as a non-keyword, in which case the behavior may be silently changing?

(if backward compat werent an issue, I'd be OK with making everything other than `n` and `normalize` keyword-only in all these classes)

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
+            raise TypeError("Offset received both month and startingMonth")
+        elif month is None and startingMonth is not None:
+            warnings.warn(
+                "startingMonth is deprecated, use month instead.",
```
should this explain that the meaning has changed?  or point to a link?

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
+            warnings.warn(
+                "startingMonth is deprecated, use month instead.",
+                FutureWarning,
+                stacklevel=2
```
trailing comma, pretend we could run `black` on this file

Pawel-Kranzberg replied:
Fixed, hopefully.

Pawel-Kranzberg committed:
Update v1.3.0.rst

Pawel-Kranzberg committed:
Update offsets.pyx

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
 
+.. _whatsnew_130.notable_bug_fixes.quarterbegin_default_anchor:
+
+Changed default periods in :class:`~pandas.tseries.offsets.QuarterBegin` and :class:`~pandas.tseries.offsets.BQuarterBegin`
```
this needs to be the same length as the title, it should error if not

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
+
+*pandas 1.3.0*
+
+.. code-block:: ipython
```
don't use a code-block, rather an ipython block for the new behavior

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
     df.groupby('label1').rolling(1).sum()
 
 
+.. _whatsnew_130.notable_bug_fixes.quarterbegin_default_anchor:
```
you need to say deprecation about the `startingMonth` as its mentioned below as well. I would prefer making these 2 distinct and unrelated changes (which they are). so don't mention deprecation at all here.

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -625,6 +656,7 @@ Datetimelike
 - Bug in :meth:`Timedelta.round`, :meth:`Timedelta.floor`, :meth:`Timedelta.ceil` for values near the implementation bounds of :class:`Timedelta` (:issue:`38964`)
 - Bug in :func:`date_range` incorrectly creating :class:`DatetimeIndex` containing ``NaT`` instead of raising ``OutOfBoundsDatetime`` in corner cases (:issue:`24124`)
 - Bug in :func:`infer_freq` incorrectly fails to infer 'H' frequency of :class:`DatetimeIndex` if the latter has a timezone and crosses DST boundaries (:issue:`39556`)
+- Bug in offsets :class:`QuarterBegin` and :class:`BQuarterBegin` returning days that are not conventional quarter beginnings (:issue:`8435`)
```
isn't this a duplicate of the above section?

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
-            startingMonth = self._default_starting_month
-        self.startingMonth = startingMonth
+        # GH#5307 startingMonth retained for backwards compatibility
+        if month and startingMonth and month != startingMonth:
```
don't you need to check not-none? e.g. 0 is value that would hit this too (which would fail but is misleading no?)

jreback reviewed (changes requested):
@Pawel-Kranzberg can you split this change up. There is just too much going on here.

- doc-strings
- deprecation
- breaking change
-
I would *really* like to do the deprecation first and separate

simonjayhawkins commented:
> I would _really_ like to do the deprecation first and separate

@Pawel-Kranzberg what's the status. close this and open separate PRs?

Pawel-Kranzberg commented:
> > I would _really_ like to do the deprecation first and separate
> 
> @Pawel-Kranzberg what's the status. close this and open separate PRs?

Yes, I plan to do so next week.

mroeschke commented:
Thanks for the PR, but it has appeared to have gone stale. Closing due to inactivity, but we would be happy to take a new PR working on the deprecation first.
